tools/chainsync-loadtest/src/network.rs

```diff
@@ -0,0 +1,330 @@
+        }
+    }
+    .parse()
+    .unwrap();
+}
+
+#[derive(Default, Debug)]
+pub struct Stats {
+    msgs_sent: AtomicU64,
+    msgs_recv: AtomicU64,
+
+    header_start: AtomicU64,
+    header_done: AtomicU64,
+    block_start: AtomicU64,
+    block_done: AtomicU64,
+    chunk_start: AtomicU64,
+    chunk_done: AtomicU64,
+}
+
+struct NetworkData {
+    info_futures: Vec<oneshot::Sender<Arc<NetworkInfo>>>,
+    info_: Arc<NetworkInfo>,
+}
+
+pub struct Network {
+    pub stats: Stats,
+    network_adapter: Arc<dyn PeerManagerAdapter>,
+    pub block_headers: Arc<WeakMap<CryptoHash, Once<Vec<BlockHeader>>>>,
+    pub blocks: Arc<WeakMap<CryptoHash, Once<Block>>>,
+    pub chunks: Arc<WeakMap<ChunkHash, Once<PartialEncodedChunkResponseMsg>>>,
+    data: Mutex<NetworkData>,
+
+    chain_id: String,
+    // client_config.min_num_peers
+    min_peers: usize,
+    // Currently it is equivalent to genesis_config.num_block_producer_seats,
+    // (see https://cs.github.com/near/nearcore/blob/dae9553670de13c279d3ebd55f17da13d94fa691/nearcore/src/runtime/mod.rs#L1114).
+    // AFAICT eventually it will change dynamically (I guess it will be provided in the Block).
+    parts_per_chunk: u64,
+
+    request_timeout: tokio::time::Duration,
+    rate_limiter: RateLimiter,
+}
+
+impl Network {
+    pub fn new(
+        config: &NearConfig,
+        network_adapter: Arc<dyn PeerManagerAdapter>,
+        qps_limit: u32,
+    ) -> Arc<Network> {
+        Arc::new(Network {
+            stats: Default::default(),
+            network_adapter,
+            data: Mutex::new(NetworkData {
+                info_: Arc::new(NetworkInfo {
+                    connected_peers: vec![],
+                    num_connected_peers: 0,
+                    peer_max_count: 0,
+                    highest_height_peers: vec![],
+                    sent_bytes_per_sec: 0,
+                    received_bytes_per_sec: 0,
+                    known_producers: vec![],
+                    peer_counter: 0,
+                }),
+                info_futures: Default::default(),
+            }),
+            blocks: WeakMap::new(),
+            block_headers: WeakMap::new(),
+            chunks: WeakMap::new(),
+
+            chain_id: config.client_config.chain_id.clone(),
+            min_peers: config.client_config.min_num_peers,
+            parts_per_chunk: config.genesis.config.num_block_producer_seats,
+            rate_limiter: RateLimiter::new(
+                time::Duration::from_secs(1) / qps_limit,
+                qps_limit as u64,
+            ),
+            request_timeout: time::Duration::from_secs(2),
+        })
+    }
+
+    fn keep_sending(
+        self: &Arc<Self>,
+        ctx: &Ctx,
+        new_req: impl Fn(FullPeerInfo) -> NetworkRequests + Send,
+    ) -> impl Future<Output = anyhow::Result<()>> + Send {
+        let self_ = self.clone();
+        let ctx = ctx.with_label("keep_sending");
+        async move {
+            loop {
```

Comment:
when do we exit this loop?

Reply:
Only if the context expires.
The happy path looks as follows:
1. we spawn keep_sending()
2. we await a response
3. once we get the response, we cancel keep_sending(), by cancelling the context.